pkg/mco/mcpList.go

```diff
@@ -8,83 +8,73 @@ import (
 
-	mcov1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/util/wait"
 )
 
-// MCPListBuilder provides struct for MachineConfigPoolList object which contains connection to cluster
-// and MachineConfigPoolList definitions.
-type MCPListBuilder struct {
-	// MachineConfigPoolList definition. Used to create MachineConfigPoolList object with minimum
-	// set of required elements.
-	ObjectList *mcov1.MachineConfigPoolList
-	apiClient  *clients.Settings
-	// machineConfigSelector label.
-	mcSelector string
-}
-
-// NewMCPListBuilder method creates new instance of MCPListBuilder.
-func NewMCPListBuilder(apiClient *clients.Settings, mcpLabelSelector ...map[string]string) *MCPListBuilder {
-	glog.V(100).Infof(
-		"Initializing new MCPListBuilder structure with the following params: %v", mcpLabelSelector)
-
-	builder := &MCPListBuilder{
-		apiClient: apiClient,
-	}
+func ListMCP(apiClient *clients.Settings, mcpLabelSelector ...map[string]string) ([]*MCPBuilder, error) {
+	listOptions := metav1.ListOptions{}
 
 	if len(mcpLabelSelector) > 0 {
-		// Serialize selector
-		serialSelector := labels.Set(mcpLabelSelector[0]).String()
-		builder.mcSelector = serialSelector
-		glog.V(100).Infof("NewMCPListBuilder builder.mcSelector is: %v", builder.mcSelector)
+		listOptions.LabelSelector = labels.Set(mcpLabelSelector[0]).String()
 	}
 
-	return builder
-}
+	glog.V(100).Infof("Listing all MCP resources with the options %v\"", listOptions)
 
-// Discover method gets the MachineConfigPools in cluster and stores them in the builder struct.
-func (builder *MCPListBuilder) Discover() error {
-	glog.V(100).Infof("Getting the list of MachineConfigPool objects on this cluster")
+	mcpList, err := apiClient.MachineConfigPools().List(context.Background(), listOptions)
 
-	var (
-		mcpList *mcov1.MachineConfigPoolList
-		err     error
-	)
+	if err != nil {
+		glog.V(100).Infof("Failed to list MCP objects due to %s", err.Error())
 
-	if builder.mcSelector != "" {
-		mcpList, err = builder.apiClient.MachineConfigPools().List(
-			context.TODO(), metav1.ListOptions{LabelSelector: builder.mcSelector})
-	} else {
-		mcpList, err = builder.apiClient.MachineConfigPools().List(
-			context.TODO(), metav1.ListOptions{})
+		return nil, err
 	}
 
-	if err != nil {
-		glog.V(100).Infof("Error to list MachineConfigPools")
+	var mcpObjects []*MCPBuilder
+
+	for _, mcp := range mcpList.Items {
+		copiedMcp := mcp
+		mcpBuilder := &MCPBuilder{
+			apiClient:  apiClient,
+			Object:     &copiedMcp,
+			Definition: &copiedMcp,
+		}
 
-		return err
+		mcpObjects = append(mcpObjects, mcpBuilder)
 	}
 
-	if len(mcpList.Items) < 1 {
-		glog.V(100).Infof("Cluster doesn't have MachineConfigPools installed ")
+	return mcpObjects, nil
+}
+
+func GetByLabelFromMCPList(apiClient *clients.Settings, mcpLabel string) (*MCPBuilder, error) {
```

Comment:
This function seems to not only check labels but also expressions. Maybe it would be better to pass in a label selector and evaluate the machineconfigpool's Spec.MachineConfigSelector.MatchExpressions and Spec.MachineConfigSelector.MatchLabels values against it?